src/main/java/net/neoforged/neoforge/common/util/AttributeUtil.java

```diff
@@ -0,0 +1,356 @@
+import net.minecraft.core.component.DataComponents;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.network.chat.CommonComponents;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.TextColor;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.effect.MobEffect;
+import net.minecraft.world.entity.EquipmentSlotGroup;
+import net.minecraft.world.entity.ai.attributes.Attribute;
+import net.minecraft.world.entity.ai.attributes.AttributeModifier;
+import net.minecraft.world.entity.ai.attributes.AttributeModifier.Operation;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.TooltipFlag;
+import net.minecraft.world.item.alchemy.PotionContents;
+import net.minecraft.world.item.component.ItemAttributeModifiers;
+import net.neoforged.fml.loading.FMLEnvironment;
+import net.neoforged.neoforge.client.event.AddAttributeTooltipsEvent;
+import net.neoforged.neoforge.client.event.GatherSkippedAttributeTooltipsEvent;
+import net.neoforged.neoforge.common.NeoForge;
+import net.neoforged.neoforge.common.NeoForgeMod;
+import net.neoforged.neoforge.common.extensions.IAttributeExtension;
+import net.neoforged.neoforge.event.ItemAttributeModifierEvent;
+import net.neoforged.neoforge.internal.versions.neoforge.NeoForgeVersion;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Utility code to support {@link IAttributeExtension}.
+ */
+public class AttributeUtil {
+    /**
+     * ID of the base modifier for Attack Damage
+     */
+    public static final ResourceLocation BASE_ATTACK_DAMAGE_ID = Item.BASE_ATTACK_DAMAGE_ID;
+
+    /**
+     * ID of the base modifier for Attack Speed
+     */
+    public static final ResourceLocation BASE_ATTACK_SPEED_ID = Item.BASE_ATTACK_SPEED_ID;
+
+    /**
+     * ID of the base modifier for Attack Range
+     */
+    public static final ResourceLocation BASE_ENTITY_REACH_ID = ResourceLocation.withDefaultNamespace("base_entity_reach");
+
+    /**
+     * ID used for attribute modifiers used to hold merged values when {@link NeoForgeMod#enableMergedAttributeTooltips()} is active.
+     * <p>
+     * Should not be used by any real attribute modifiers for gameplay purposes.
+     */
+    public static final ResourceLocation FAKE_MERGED_ID = ResourceLocation.fromNamespaceAndPath(NeoForgeVersion.MOD_ID, "fake_merged_modifier");
+
+    /**
+     * Comparator for {@link AttributeModifier}. First compares by operation, then amount, then the ID.
+     */
+    public static final Comparator<AttributeModifier> ATTRIBUTE_MODIFIER_COMPARATOR = Comparator.comparing(AttributeModifier::operation)
+            .thenComparing(Comparator.comparingDouble(a -> -Math.abs(a.amount()))) // Sort most impactful modifiers first
+            .thenComparing(Comparator.comparing(AttributeModifier::id));
```

Comment:
Please replace the `.thenComparing(Comparator.comparingDouble(` with `.thenComparingDouble(`.

And remove the unnecessary `Comparator.comparing` inside the second `thenComparing` call, unless we potentially should be passing an extra parameter to that call so that it compares by `ResourceLocation#compareNamespaced` which might make sense for us to do.

Reply:
I think using `ResourceLocation#compareTo` is more correct here, because it will produce a (probably) alphabetically-sorted list of attributes by path, rather than invoking the namespace as part of the equation.  I'll clean up the `thenComparing` stuff though.